### ISSUE TITLE
Import couldn't be resolved when a subdirectory package contains a pyproject.toml file

### ISSUE THREAD

**Issue by nelsyeung** [microsoft/pyright#4039]
**Describe the bug**
Import could not be resolved when a subdirectory package is installed that contains a pyproject.toml file.

**To Reproduce**
1. Just run these shell commands to create all the necessary files:

    ```sh
    mkdir -p foo/bar/bar foo/src
    
    cat <<EOT >> foo/bar/pyproject.toml
    [build-system]
    requires = ["setuptools >= 61.0.0"]
    build-backend = "setuptools.build_meta"
    
    [project]
    name = "bar"
    description = "Bar"
    requires-python = ">=3.7"
    version = "1.0.0"
    EOT
    
    touch foo/bar/bar/__init__.py
    echo 'bar = "bar"' > foo/bar/bar/consts.py
    echo "from bar.consts import bar; print(bar)" > foo/src/foo.py
    
    cd foo
    
    python -m venv .venv
    source .venv/bin/activate
    
    pip install -U pip setuptools
    pip install -e ./bar
    ```
1. Open the `src/foo.py` file, you should see the error:

    ```
    Import "bar.consts" could not be resolved (Pyright reportMissingImports)
    ```

1. However, running the script `python src/foo.py` is perfectly fine.

**Expected behavior**
Resolves correctly

**Screenshots or Code**

<img width="601" alt="Screenshot 2022-10-12 at 12 26 11" src="https://user-images.githubusercontent.com/3951541/195330908-5b19a9c7-6359-47d0-961d-7c3f7384be5c.png">

**VS Code extension or command-line**
This happens on both VS Code Pylance and coc-pyright.
pyright version: `1.1.272`

**Additional context**
If the `pyproject.toml` file is not there, it's fine. That is, if the package is built using `setup.py` it's fine, but note that you still can't have `pyproject.toml` file there even when used with `setup.py`. Basically, pyright recognise the `pyproject.toml` inside the `foo/bar` folder as a module instead of `foo/bar/bar`, so when it resolves `import bar`, it's looking at `foo/bar` and not `foo/bar/bar`.

**Comment by erictraut** [comment]
I'm seeing the same behavior regardless of whether the `pyproject.toml` file is present. The same import resolution error is reported in both cases.

For a documentation on how pyright resolves imports, refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/import-resolution.md).

In your case, you can fix the problem by telling pyright that you want it to explore `bar` for imports. You can do that by adding a `pyrightconfig.json` file at the top level of your project.

```json
{
    "extraPaths": ["bar"]
}
```

Or you can add a second pyproject.toml that contains the equivalent.

**Comment by nelsyeung** [comment]
@erictraut  I'm not sure how you tried to reproduce this, I'm not seeing the error when I don't use a `pyproject.toml`. Have you created a `setup.py` instead? Have you cleaned your environment? Use this script here to see what I mean by working without `pyproject.toml`:

```sh
mkdir -p foo/bar/bar foo/src

cat <<EOT >> foo/bar/setup.py
from setuptools import find_packages, setup

setup(
    name="bar",
    version="1.0.0",
    packages=find_packages(),
)
EOT

touch foo/bar/bar/__init__.py
echo 'bar = "bar"' > foo/bar/bar/consts.py
echo "from bar.consts import bar; print(bar)" > foo/src/foo.py

cd foo

python -m venv .venv
source .venv/bin/activate

pip install -U pip setuptools
pip install -e ./bar
```

In addition, this used to work as expected about 2 weeks ago (although I can't confirm the date).

**Comment by nelsyeung** [comment]
Just in case you missed it, we have installed the package `bar` (i.e., `pip install -e ./bar`), so pyright not resolving correctly and requires manual intervention just makes no sense to me.

**Comment by erictraut** [comment]
To repro, I ran your original script. I'll re-open and take another look at the updated script you provided.

**Comment by debonte** [comment]
When using your `pyproject.toml`, setuptools is implementing the editable install using import hooks.

```console
> more foo\.venv\Lib\site-packages\__editable__.bar-1.0.0.pth
import __editable___bar_1_0_0_finder; __editable___bar_1_0_0_finder.install()
```

When using your `setup.py` instead, it uses a simple `.pth` file. I'm unsure why the behavior of setuptools differs in the two cases.

Pyright can't resolve imports using import hooks. There are more details including workarounds [here](https://github.com/microsoft/pyright/blob/main/docs/import-resolution.md#editable-installs).

**Comment by erictraut** [comment]
Ah, thanks for noticing that @debonte. Yes, that explains the problem.

@nelsyeung, please follow the instructions in the linked documentation. As Erik said, static analysis tools like pyright will not work with dynamic import hooks. Unfortunately, the latest version of setuptools makes that the default for editable installs. This is causing significant pain for developers as we predicted it would. You may want to provide feedback to the setuptools maintainers about this unfortunate decision.
